### PR TITLE
[DBFTPlugin] Dbft constructor

### PR DESCRIPTION
**Pull request by ZhangTao1596**
Close https://github.com/neo-project/neo-modules/issues/455 https://github.com/neo-project/neo-modules/issues/543 https://github.com/neo-project/neo-modules/issues/544

Changes:
 * Avoid using static
 * Provide settings in constructors of `ConsensusService` and `ConsensusContext`.

@erikzhang @shargon @devhawk

### CODE REVIEW

File: src/DBFTPlugin/ConsensusMessage.cs [hunk 1]
```diff
@@ -6,6 +6,7 @@ namespace Neo.Consensus
 {
     public abstract class ConsensusMessage : ISerializable
     {
+        protected readonly byte validatorsCount;
```
Comment (erikzhang) [inline src/DBFTPlugin/ConsensusMessage.cs:9]: Why do we need this?

Comment (ZhangTao1596) [reply src/DBFTPlugin/ConsensusMessage.cs:9]: For index check in deserialize.
I removed `DBFTPlugin.System`.

Comment (erikzhang) [reply src/DBFTPlugin/ConsensusMessage.cs:9]: Maybe we can check it in `Verify()`?

Comment (ZhangTao1596) [reply src/DBFTPlugin/ConsensusMessage.cs:9]: Will do. It'll looks better.

Comment (shargon) [reply src/DBFTPlugin/ConsensusMessage.cs]: Changed except `RecoveryMessage`, please take a look

**Review by shargon — APPROVED** [review]

**Review by shargon — APPROVED** [review]
Wait for @superboyiii tests

**Review by devhawk — CHANGES_REQUESTED** [review]
Settings.Default should be eliminated

File: src/DBFTPlugin/DBFTPlugin.cs [hunk 2]
```diff
@@ -23,16 +23,16 @@ protected override void Configure()
         protected override void OnSystemLoaded(NeoSystem system)
         {
             if (system.Settings.Magic != Settings.Default.Network) return;
```
[the remark this answers is not in the document]
Comment (ZhangTao1596) [reply src/DBFTPlugin/DBFTPlugin.cs]: You can construct `ConsensusContext` with your own settings. Why you need `DBFTPlugin`?

File: src/DBFTPlugin/ConsensusMessage.cs [hunk 3]
```diff
@@ -30,18 +30,29 @@ public virtual void Deserialize(BinaryReader reader)
...
+            {
+                ConsensusMessageType.PrepareRequest => new PrepareRequest(),
+                ConsensusMessageType.PrepareResponse => new PrepareResponse(),
+                ConsensusMessageType.ChangeView => new ChangeView(),
+                ConsensusMessageType.Commit => new Commit(),
+                ConsensusMessageType.RecoveryRequest => new RecoveryRequest(),
+                ConsensusMessageType.RecoveryMessage => new RecoveryMessage(validatorsCount),
+                _ => throw new FormatException(),
+            };
+            message.Deserialize(reader);
```
Comment (erikzhang) [inline src/DBFTPlugin/ConsensusMessage.cs:52]: I don't think it works. You have read the `ConsensusMessageType` once before, and you will read it again in `Deserialize()`.

Comment (ZhangTao1596) [reply src/DBFTPlugin/ConsensusMessage.cs:52]: Done.

File: src/DBFTPlugin/DBFTPlugin.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (devhawk) [reply src/DBFTPlugin/DBFTPlugin.cs]: neo express loads DBFTPlugin statically when running a local privatenet node. The magic value comes from the neo-express file *NOT* from a plugin config file.

File: src/DBFTPlugin/DBFTPlugin.cs [hunk 4]
```diff
@@ -56,7 +57,7 @@ public void Start(Wallet wallet)
         {
```
Comment (devhawk) [inline src/DBFTPlugin/DBFTPlugin.cs:57]: Can we add an optional Settings parameter to the Start method that overrides whatever was loaded by Configure?

**Review by devhawk — CHANGES_REQUESTED** [review]
I still need a mechanism to pass custom settings to `DBFTPlugin`. I suggested adding an optional settings parameter to `DBFTPlugin.Start` method

File: src/DBFTPlugin/DBFTPlugin.cs (hunk 4 above)
Comment (devhawk) [reply src/DBFTPlugin/DBFTPlugin.cs:57]: Suggestion:

```
        public void Start(Wallet wallet, Settings settings = null)
        {
            if (started) return;
            if (settings != null) this.settings = settings;
            started = true;
            consensus = neoSystem.ActorSystem.ActorOf(ConsensusService.Props(neoSystem, this.settings, wallet));
            consensus.Tell(new ConsensusService.Start());
        }
```

Comment (ZhangTao1596) [reply src/DBFTPlugin/DBFTPlugin.cs:57]: Done.

**Comment by superboyiii** [comment]
Test **PASS**
Works well on mine.
image

**Review by superboyiii — DISMISSED** [review]

**Comment by superboyiii** [comment]
@erikzhang @devhawk Need your review again.

File: src/DBFTPlugin/DBFTPlugin.cs [hunk 5]
```diff
@@ -53,10 +53,11 @@ private void OnStart()
             Start(walletProvider.GetWallet());
         }
 
-        public void Start(Wallet wallet)
+        public void Start(Wallet wallet, Settings settings = null)
         {
             if (started) return;
             started = true;
+            if (settings != null) this.settings = settings;
             consensus = neoSystem.ActorSystem.ActorOf(ConsensusService.Props(neoSystem, settings, wallet));
```
Comment (devhawk) [inline src/DBFTPlugin/DBFTPlugin.cs:61]: you're missing a `this` on line 60. when there's a method parameter and class field with the same identitifer (aka `settings`) you need to use `this.` to distinguish between them.  The way the code is now, the settings passed to `ConsensusService.Props` will always be the parameter `settings`, even if it's null.

``` csharp
consensus = neoSystem.ActorSystem.ActorOf(ConsensusService.Props(neoSystem, this.settings, wallet));
```

**Review by devhawk — COMMENTED** [review]
Fix the missing `this.` in Start and it looks good to me

**Review by devhawk — DISMISSED** [review]

File: src/DBFTPlugin/RecoveryMessage.cs [hunk 6]
```diff
@@ -79,7 +92,7 @@ internal ExtensiblePayload[] GetPrepareResponsePayloads(ConsensusContext context
         {
             UInt256 preparationHash = PreparationHash ?? context.PreparationPayloads[context.Block.PrimaryIndex]?.Hash;
             if (preparationHash is null) return Array.Empty<ExtensiblePayload>();
-            return PreparationMessages.Values.Where(p => p.ValidatorIndex != context.Block.PrimaryIndex).Select(p => context.CreatePayload(new PrepareResponse
+            return PreparationMessages.Values.Where(p => p.ValidatorIndex != context.Block.PrimaryIndex).Select(p => context.CreatePayload(new PrepareResponse()
```
Comment (erikzhang) [inline src/DBFTPlugin/RecoveryMessage.cs:95]: Can you revert this kind of changes?

File: src/DBFTPlugin/RecoveryMessage.cs [hunk 7]
```diff
@@ -17,32 +17,45 @@ public partial class RecoveryMessage : ConsensusMessage
...
+
         public override int Size => base.Size
             + /* ChangeViewMessages */ ChangeViewMessages?.Values.GetVarSize() ?? 0
             + /* PrepareRequestMessage */ 1 + PrepareRequestMessage?.Size ?? 0
             + /* PreparationHash */ PreparationHash?.Size ?? 0
             + /* PreparationMessages */ PreparationMessages?.Values.GetVarSize() ?? 0
             + /* CommitMessages */ CommitMessages?.Values.GetVarSize() ?? 0;
 
-        public RecoveryMessage() : base(ConsensusMessageType.RecoveryMessage)
+        public RecoveryMessage(int validatorsCount) : base(ConsensusMessageType.RecoveryMessage)
```
Comment (erikzhang) [inline src/DBFTPlugin/RecoveryMessage.cs:29]: Why do we need `validatorsCount` in the constructor?

Comment (shargon) [reply src/DBFTPlugin/RecoveryMessage.cs:29]: It's composed for different messages that we need to validate

**Review by devhawk — CHANGES_REQUESTED** [review]
still waiting on the `this.` fix in `Start`

File: src/DBFTPlugin/RecoveryMessage.cs (hunk 7 above)
Comment (erikzhang) [reply src/DBFTPlugin/RecoveryMessage.cs:29]: Can we verify it in `Verify()` method?

**Comment by ZhangTao1596** [comment]
> still waiting on the `this.` fix in `Start`

Done.

File: src/DBFTPlugin/RecoveryMessage.cs (hunk 7 above)
Comment (ZhangTao1596) [reply src/DBFTPlugin/RecoveryMessage.cs]: Done.

**Review by shargon — COMMENTED** [review]
Tested?

File: src/DBFTPlugin/RecoveryMessage.cs [hunk 8]
```diff
@@ -24,25 +24,34 @@ public partial class RecoveryMessage : ConsensusMessage
...
-        public RecoveryMessage() : base(ConsensusMessageType.RecoveryMessage)
-        {
-        }
+        public RecoveryMessage() : base(ConsensusMessageType.RecoveryMessage) { }
 
         public override void Deserialize(BinaryReader reader)
         {
             base.Deserialize(reader);
-            ChangeViewMessages = reader.ReadSerializableArray<ChangeViewPayloadCompact>(DBFTPlugin.System.Settings.ValidatorsCount).ToDictionary(p => (int)p.ValidatorIndex);
+            ChangeViewMessages = reader.ReadSerializableArray<ChangeViewPayloadCompact>(ushort.MaxValue).ToDictionary(p => (int)p.ValidatorIndex);
```
Comment (erikzhang) [inline src/DBFTPlugin/RecoveryMessage.cs:32]: I think `byte` is enough.

Comment (ZhangTao1596) [reply src/DBFTPlugin/RecoveryMessage.cs]: Done.

**Review by erikzhang — APPROVED** [review]

**Comment by erikzhang** [comment]
@devhawk Can you review it again?

**Review by devhawk — APPROVED** [review]

**Review by shargon — APPROVED** [review]

**Review by erikzhang — APPROVED** [review]

**Comment by erikzhang** [comment]
Merge?

**Pull request merged**